narwhal/worker/src/batch_maker.rs

```diff
@@ -115,26 +115,32 @@ impl BatchMaker {
                 // Note that transactions are only consumed when the number of batches
                 // 'in-flight' are below a certain number (MAX_PARALLEL_BATCH). This
                 // condition will be met eventually if the store and network are functioning.
-                Some((transaction, response_sender)) = self.rx_batch_maker.recv(), if batch_pipeline.len() < MAX_PARALLEL_BATCH => {
+                Some((transactions, response_sender)) = self.rx_batch_maker.recv(), if batch_pipeline.len() < MAX_PARALLEL_BATCH => {
                     let _scope = monitored_scope("BatchMaker::recv");
-                    current_batch_size += transaction.len();
-                    current_batch.transactions_mut().push(transaction);
+
+                    // If there are multiple transactions, we only send back the digest of the first Batch.
+                    // This currently poses no issue but needs to be fixed should a caller need to know the digest of all batches.
                     current_responses.push(response_sender);
-                    if current_batch_size >= self.batch_size_limit {
-                        if let Some(seal) = self.seal(false, current_batch, current_batch_size, current_responses).await{
-                            batch_pipeline.push(seal);
-                        }
-                        self.node_metrics.parallel_worker_batches.set(batch_pipeline.len() as i64);
+                    for transaction in transactions {
+                        current_batch_size += transaction.len();
+                        current_batch.transactions_mut().push(transaction);
 
-                        current_batch = Batch::new(vec![], &self.protocol_config);
-                        current_responses = Vec::new();
-                        current_batch_size = 0;
+                        if current_batch_size >= self.batch_size_limit {
+                            if let Some(seal) = self.seal(false, current_batch, current_batch_size, current_responses).await{
```

Comment:
Just to reconfirm, based on the [SIP-19](https://github.com/sui-foundation/sips/pull/19/files#diff-d2a12d680ca632ffe5172eedfc6075e46698eb221f61158f7e0e5a8c86aa7e53R60) `Note that the Soft Bundle API does not provide a strict ordering guarantee. However, it does provide an extremely high probability of being ordered correctly.`  , the intention is best effort to fit multiple transactions in the same batch, right ?

Also how do we intend to measure the successful inclusion of the soft bundle transactions in same batch/block? I assume the appearance in checkpoints would be one way if lower level metrics don't make much sense

Reply:
> the intention is best effort to fit multiple transactions in the same batch

That's right!  The intention is to include the txs as a bundle if feasible to do so.  The pre-existing constraints (e.g. batch size limit) are there for a good reason, so that should come first.   It is alright for a Soft Bundle to fail.

> I assume the appearance in checkpoints would be one way if lower level metrics don't make much sense

Exactly.  It seems natural for the client to monitor the bundle it submitted, by looking at whether transactions are ordered as specified in the eventual checkpoint.  If the bundles are failing frequently, one can tell that something must be wrong.